Documentation/ApiOverview/DependencyInjection/Index.rst

```diff
@@ -12,59 +16,466 @@ Dependency injection
+this cache has to be manually emptied for a running instance by clicking this button.
+The backend extension manager however *does* empty the cache automatically when loading or unloading extensions.
+Another way to quickly drop this cache during development is to remove all
+:file:`var/cache/code/di/*` files, which reside in :file:`typo3temp/` in non-composer based
+instances or elsewhere in composer based instances (see :ref:`Environment`). TYPO3 will
+then recalculate the cache upon the next instance call, no matter if it's a frontend, a backend
+or a CLI request.
+
+The main takeaway is: When a developer fiddles with container configuration,
+the cache needs to be manually cleared. And if some configuration issue slipped in,
+which made the container or DI calculation fail, the system does *not* heal itself and
+needs both a fix of the configuration plus probably a cache removal. The standalone install
+tool however should *always* work, even if the backend breaks down, so the "Flush caches"
+button is always reachable. Note that *if* the container calculation fails, the
+:file:`var/log/typo3_*` files contain the exception with backtrace!
+
+
+Important terms
+===============
+
+We will continue to use a couple of technical terms in this chapter, so let's quickly
+define them to align. Some of them are not precisely used in our world, for
+instance some Java devs may stumble upon "our" understanding of a prototype.
+In general, we're still somehow using them in the right direction ;)
+
+Prototype
+   The broad understanding of a *prototype* within the TYPO3 community is
+   that it's simply an object that is created a-new every time. Basically the direct
+   opposite of a *singleton*. In fact, the prototype pattern describes a base object that
+   is created once, so :php:`__construct()` is called to set it up, after that it is
+   cloned each time one wants to have a new instance of it. The community isn't well aware
+   of that, and the core provides no "correct" prototype API, so the word *prototype* is
+   often misused for an object that is always created a-new when the framework is
+   asked to create one.
+
+Singleton
+   That's pretty clear. A *singleton* is an object that is instantiated
+   exactly once within one process. If an instance is requested and the object has been
+   created already once, the same instance is returned. Codewise, this is sometimes done by
+   implementing a static :php:`getInstance()` method that parks the instance in a property.
+   In TYPO3, this can also be achieved by implementing the :php:`SingletonInterface`,
+   where :php:`makeInstance()` then stores the object internally. Within containers, this can be done
+   by declaring the object :yaml:`shared: true`, which is usually the default. We'll come back to
+   details later. Singletons *must not* have state - they must act the same way each time
+   they're used, no matter where, how often or when they've been used before.
```

Comment:
```suggestion
   they're used, no matter where, how often or when they've been used before. Otherwise the behavior of a singleton
   object is undefined and will lead to obscure errors.
```